### PR TITLE
Fuse per-sequence AlltoAll into a unified one in GDN forward

### PR DESCRIPTION
**Pull request by xuantengh**
# What does this PR do ?

PR #2645 adds the THD support for GDN, but it requires per-sequence, per-section (qkv, alpha, beta, gate) AlltoAll converting activation between CP <-> HP. This PR fuses the per-sequence AlltoAll into a unified one, with local indexing mappings to reorder at the sequence dimension and feature dimension.

[wandb link](https://wandb.ai/nvidia/mcore-qwen3-gdn?nw=nwuserxuantengh) for running a Qwen3 Next proxy model with 16K sequence length and CP size = 4, the iteration time speedup is around 7% - 10%.

image

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by xuantengh** [comment]
/ok to test [8dcfd7d](https://github.com/NVIDIA/Megatron-LM/pull/4913/commits/8dcfd7d90497420b5669c05a3a91e0d04eae933e)

**Comment by xuantengh** [comment]
/ok to test [6d28256](https://github.com/NVIDIA/Megatron-LM/pull/4913/commits/6d2825677513ef89a7fde998d4a91805f49f3f50)

**Comment by Phlip79** [comment]
What is the performance impact of this change?

**Comment by xuantengh** [comment]
> What is the performance impact of this change?

Hi Philip, I've ran a proxy model of Qwen3 Next with 16K seq length and CP size set to 4. The iteration time speed up is around 7% - 10%, see the PR description.

Note that this was under the `sbhd` input layout. For `thd` sequence packing scenario, the speedup will be higher theoretically, as it fuses the per-sequence AlltoAll. Currently there is no recipe to run `thd`, we're waiting for #4832 for this support.

File: megatron/core/ssm/gated_delta_net.py [hunk 1]
```diff
@@ -708,6 +717,63 @@ def _unpack_sequence(x, cu_seqlens, dim=1):
...
+    cu_seqlens: torch.Tensor, cp_size: int, t_global: int
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    cu = cu_seqlens.to(dtype=torch.long)
+    t_local = t_global // cp_size
+
+    positions = torch.arange(t_global, device=cu.device)
+    seq_idx = torch.bucketize(positions, cu[1:], right=True)
+    seq_lens = torch.diff(cu)
+    halves = seq_lens // (2 * cp_size)  # per-sequence half-chunk size
+    local_starts = cu[:-1] // cp_size
```
[the remark this answers is not in the document]
Comment (xuantengh) [reply megatron/core/ssm/gated_delta_net.py:728]: CP load balance already requires each sequence length divisible by 2 * `cp_size`.

**Comment by xuantengh** [comment]
/ok to test [d2872d7](https://github.com/NVIDIA/Megatron-LM/pull/4913/commits/d2872d74565328fc9ac5e610a6dc01433a2ff1c2)

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/ssm/gated_delta_net.py [hunk 2]
```diff
@@ -138,6 +138,24 @@ def __init__(
...
+                    self.num_value_heads // self.tp_size,
+                    self.num_value_heads // self.tp_size,
+                ],
+                self.cp_size,
+                torch.cuda.current_device(),
+            )
+        else:
+            head_perm = None
+        # Registered as a non-persistent buffer to exclude it from state_dict
+        self.register_buffer("_thd_head_perm", head_perm, persistent=False)
```
Comment (yuzhongw-nvidia) [inline megatron/core/ssm/gated_delta_net.py]: Hi @xuantengh , do you think it is a good idea to have an @lru_cache decorated method for the _thd_head_perm in each forward step, instead of straightforwardly pre-processing it in __init__?
We're going to support dynamic CP for GDN soon, where CP size varies across micro-batches (code). In that case, directly pre-processing the _thd_head_perm for a static CP size does not work.

Comment (xuantengh) [reply megatron/core/ssm/gated_delta_net.py]: Resolved, thanks.

File: megatron/core/ssm/gated_delta_net.py [hunk 3]
```diff
@@ -696,16 +697,61 @@ def _backward_out_proj(self):
...
+
+    idx = rank * t_local + local_starts[seq_idx] + k
+
+    inv = torch.empty_like(idx)
+    inv[idx] = positions
+
+    return idx, inv
+
+
+def _build_head_perm_for_split_sections(
```
Comment (yuzhongw-nvidia) [inline megatron/core/ssm/gated_delta_net.py:759]:
```suggestion
@lru_cache(maxsize = 8)
def _build_head_perm_for_split_sections(
```

Comment (xuantengh) [reply megatron/core/ssm/gated_delta_net.py:759]: Use tuple for `split_sections` to make it hashable. Thanks for your detailing review.

**Review by yuzhongw-nvidia — APPROVED** [review]
LGTM.

File: megatron/core/ssm/gated_delta_net.py [hunk 4]
```diff
@@ -696,16 +698,62 @@ def _backward_out_proj(self):
...
-    num_seqs = len(cu_seqlens_list) - 1
-    for i in range(num_seqs):
-        idx_start = cu_seqlens_list[i]
-        idx_end = cu_seqlens_list[i + 1]
-        chunked_index = [slice(None)] * dim + [slice(idx_start, idx_end)]
-        unpacked_x.append(x[tuple(chunked_index)])
-    return unpacked_x
+def _build_thd_cp_a2a_perm(
+    cu_seqlens: torch.Tensor, cp_size: int, t_global: int
+) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (janEbert) [inline megatron/core/ssm/gated_delta_net.py]: Nit: let's move toward new-style type specification using built-in `tuple` instead of `typing.Tuple`.

File: megatron/core/ssm/gated_delta_net.py [hunk 5]
```diff
@@ -696,16 +698,62 @@ def _backward_out_proj(self):
...
+
+    inv = torch.empty_like(idx)
+    inv[idx] = positions
+
+    return idx, inv
+
+
+@lru_cache(maxsize=8)
+def _build_head_perm_for_split_sections(
+    split_sections: Tuple[int], cp_size: int, device: torch.device
```
Comment (janEbert) [inline megatron/core/ssm/gated_delta_net.py]: Same nit as before, but also, the type should be `tuple[int, ...]`.

File: megatron/core/ssm/gated_delta_net.py (hunk 4 above)
Comment (xuantengh) [reply megatron/core/ssm/gated_delta_net.py]: Should I convert all type annotation (`Optional`, `Tuple`, etc.) in this file to modern one?

File: megatron/core/ssm/gated_delta_net.py (hunk 5 above)
Comment (janEbert) [reply megatron/core/ssm/gated_delta_net.py]: Type is still incorrect, should be `tuple[int, ...]`.

File: megatron/core/ssm/gated_delta_net.py (hunk 4 above)
Comment (janEbert) [reply megatron/core/ssm/gated_delta_net.py]: I would've been fine if just changed your additions, but this is nice, too. :)

File: megatron/core/ssm/gated_delta_net.py (hunk 5 above)
Comment (janEbert) [reply megatron/core/ssm/gated_delta_net.py]: Thanks!

**Review by janEbert — APPROVED** [review]

**Comment by xuantengh** [comment]
/ok to test [5d261d1](https://github.com/NVIDIA/Megatron-LM/pull/4913/commits/5d261d15eb0b2111ed96bba24127ffa82687831c)

**Comment by xuantengh** [comment]
/ok to test [4137f93](https://github.com/NVIDIA/Megatron-LM/pull/4913/commits/4137f937cf6cf2260c0afa8268536035ed7438b4)
